ecs-init/cache/cache.go

```diff
@@ -134,48 +154,69 @@ func (d *Downloader) DownloadAgent() error {
 	calculatedMd5SumString := fmt.Sprintf("%x", calculatedMd5Sum)
 	log.Debugf("Expected %s", publishedMd5Sum)
 	log.Debugf("Calculated %s", calculatedMd5SumString)
-	agentRemoteTarball := config.AgentRemoteTarball(d.getRegion())
+	agentRemoteTarball := config.AgentRemoteTarballKey()
 	if publishedMd5Sum != calculatedMd5SumString {
 		err = fmt.Errorf("mismatched md5sum while downloading %s", agentRemoteTarball)
 		return err
 	}
 
-	log.Debugf("Attempting to rename %s to %s", tempFile.Name(), config.AgentTarball())
-	return d.fs.Rename(tempFile.Name(), config.AgentTarball())
+	log.Debugf("Attempting to rename %s to %s", tempFileName, config.AgentTarball())
+	return d.fs.Rename(tempFileName, config.AgentTarball())
 }
 
 func (d *Downloader) getPublishedMd5Sum() (string, error) {
-	region := d.getRegion()
-	agentRemoteTarballMD5 := config.AgentRemoteTarballMD5(region)
-	log.Debugf("Downloading published md5sum from %s", agentRemoteTarballMD5)
-	resp, err := d.getter.Get(agentRemoteTarballMD5)
+	tempFile, err := d.fs.TempFile(config.CacheDirectory(), "ecs-agent.tar.md5")
 	if err != nil {
 		return "", err
 	}
+	log.Debugf("Temp file %s", tempFile.Name())
 	defer func() {
-		if resp != nil && resp.Body != nil {
-			resp.Body.Close()
+		if err != nil {
+			log.Debugf("Removing temp file %s", tempFile.Name())
+			d.fs.Remove(tempFile.Name())
 		}
 	}()
-	body, err := d.fs.ReadAll(resp.Body)
+	defer tempFile.Close()
+
+	_, err = d.s3downloader.Download(tempFile, &s3.GetObjectInput{
+		Bucket: aws.String(config.AgentRemoteBucketName),
+		Key: aws.String(config.AgentRemoteTarballMD5Key()),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	body, err := d.fs.ReadAll(tempFile)
 	if err != nil {
 		return "", err
 	}
+
 	return strings.TrimSpace(string(body)), nil
 }
 
-func (d *Downloader) getPublishedTarball() (io.ReadCloser, error) {
-	region := d.getRegion()
-	agentRemoteTarball := config.AgentRemoteTarball(region)
-	log.Debugf("Downloading Amazon Elastic Container Service Agent from %s", agentRemoteTarball)
-	resp, err := d.getter.Get(agentRemoteTarball)
+func (d *Downloader) getPublishedTarball() (string, error) {
+	tempFile, err := d.fs.TempFile(config.CacheDirectory(), "ecs-agent.tar")
 	if err != nil {
-		return nil, err
+		return "", err
 	}
-	if resp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("unexpected response code %d", resp.StatusCode)
+	log.Debugf("Temp file %s", tempFile.Name())
+	defer func() {
+		if err != nil {
```

Comment:
`err` gets reused multiple times in this method. Conditionally removing a file based on that can be erroneous depending on where/when `err` gets written to. Instead of doing this in `defer`, it could be less confusing to explicitly do it when this method runs into errors.

Reply:
This is still not addressed

Reply:
I added it back due to another comment.. i will remove it now